set druid.expressions.useStrictBooleans to true by default (#14734)
apache/druid: docs/querying/math-expr.md

* constants and identifiers are supported for any column type
* `cast` is supported for numeric and string types
* math operators: `+`,`-`,`*`,`/`,`%`,`^` are supported for numeric types

> * logical operators: `!`, `&&`, `||`, are supported for string and numeric types (if `druid.expressions.useStrictBooleans=true`, the default)

Reviewer: ```suggestion
* logical operators: `!`, `&&`, `||`, are supported for string and numeric types by default
```

* other: `parse_long` is supported for numeric and string types

## Logical operator modes

> In Druid 28.0 and later, `druid.expressions.useStrictBooleans=true` is set by default, so logical operations will allow correctly treating `null` values as "unknown" for SQL compatible behavior, and _all boolean output functions_ will output 'homogeneous' `LONG` typed boolean values of `1` for `true` and `0` for `false`. 

Reviewer: ```suggestion
In Druid 28.0 and later, `druid.expressions.useStrictBooleans=true` is set by default. Logical operations treat `null` values as "unknown" for SQL compatible behavior. _all boolean output functions_ output homogeneous `LONG` typed boolean values of `1` for `true` and `0` for `false`. 
```

Druid currently still retains implicit conversion of `LONG`, `DOUBLE`, and `STRING` types into boolean values in both modes: 

> * `LONG` or `DOUBLE` - any value greater than 0 is considered `true`, else `false`

Reviewer: ```suggestion
* `LONG` or `DOUBLE`: any value greater than 0 is considered `true`, else `false`.
```

> * `STRING` - the value `'true'` (case insensitive) is considered `true`, everything else is `false`.

Reviewer: ```suggestion
* `STRING`: the value `'true'` (case insensitive) is considered `true`, everything else is `false`.
```

* `'troo' && 'true'` -> `0`
* `'troo' || 'true'` -> `1`

> Prior to the 28 release of Apache Druid, boolean function expressions had inconsistent handling of true and false values, and the logical 'and' and 'or' operators behave in a manner that is incompatible with SQL, even if SQL compatible null handling mode (`druid.generic.useDefaultValueForNull=false`) is enabled. Logical operators also pass through their input values similar to many scripting languages, and treat `null` as false, which can result in some rather strange behavior. Other boolean operations, such as comparisons and equality, retain their input types (e.g. `DOUBLE` comparison would produce `1.0` for true and `0.0` for false), while many other boolean functions strictly produce `LONG` typed values of `1` for true and `0` for false. This legacy mode can be enabled by setting `druid.expressions.useStrictBooleans=false`.

Reviewer: This paragraph is confusing because you use past and present tense to refer to the past behavior. I think we should rewrite it so that it follows the same format.

> Prior to the 28 release of Apache Druid, boolean function expressions had inconsistent handling of true and false values. The logical 'and' and 'or' operators behaved in a manner that was incompatible with SQL, even if SQL compatible null handling mode (`druid.generic.useDefaultValueForNull=false`) was enabled. Logical operators would also pass through their input values similar to many scripting languages, and treated `null` as false, which would result in some rather strange behavior. Other boolean operations, such as comparisons and equality, retained their input types (e.g. `DOUBLE` comparison would produce `1.0` for true and `0.0` for false), while many other boolean functions strictly produced `LONG` typed values of `1` for true and `0` for false. This legacy mode can be enabled by setting `druid.expressions.useStrictBooleans=false`.

Reviewer 2: "Prior to Druid 28.0.0" is more consistent with how we write this kind of stuff elsewhere.

Also, I think this paragraph would read better if switched into the present tense, and written as if it's describing `druid.expressions.useStrictBooleans = false`, not as if it's describing Druid prior to 28.0.0. (Still include a mention that the default changed in 28.0.0 though.)
